Support audit ignore policy by actions

BigPandaToo opened the pull request:
Bulk operations may produce GBs of audit log events 
while most of them are related/identical. This change 
will allow to filter out events by their type of actions by 
creating a new audit ignore policy - actions.

'xpack.security.audit.logfile.events.ignore_filters:
       example1:
            actions: ["indices:data/write/bulk", "cluster:admin/ilm/*"]'

Resolve: #60877
Related: #10836
Related: #37148

Commit:
Support audit ignore policy by index privileges

Adding new audit ignore policy - privileges
For example, following policy will filter out all events, which actions
minimal required privilege is either "read" or "delete":

xpack.security.audit.logfile.events.ignore_filters:
  example:
    privileges: ["read", "delete"]

Resolve: #60877
Related: #10836
Related: #37148

elasticmachine commented:
Pinging @elastic/es-security (Team:Security)

Commit:
Support audit ignore policy by index privileges

Adding new audit ignore policy - privileges
For example, following policy will filter out all events, which actions
required privilege is either "read" or "delete":

xpack.security.audit.logfile.events.ignore_filters:
  example:
    privileges: ["read", "delete"]

Resolve: #60877
Related: #10836
Related: #37148

BigPandaToo commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into Audit_ignore_action

albertzaharovits commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/audit/logfile/LoggingAuditTrail.java:
```diff
@@ -1401,6 +1430,16 @@ private EventFilterPolicy changeIndicesFilter(List<String> filtersList) {
+            if (AuthorizationService.isIndexAction(action)) {
+                privileges = IndexPrivilege.findPrivilegesThatGrant(action);
+            } else if (ClusterPrivilegeResolver.isClusterAction(action)) {
+                privileges = ClusterPrivilegeResolver.findPrivilegesThatGrant(action, null, null);
```
I think there will be problems with the `manage_own_api_key` privilege.
In general, I dislike passing `null`s like this, it looks brittle to me.
I think if we've got the option to pass along the `authentication` and `request` parameters, we should.

albertzaharovits replied:
Is the current proposed solution that for a cluster privilege filter with `manage_own_api_keys` to ignore all audits for token actions (ie work like `manage_api_key`)?
I don't think this is the best alternative. I'd rather not permit this privilege in an ignore policy. Just my opinion, take it from someone that is not a fan of named privileges for ignore policies.

BigPandaToo replied:
Changed to use pemission check instead + added targeted tests

albertzaharovits commented on docs/reference/settings/audit-settings.asciidoc:
```diff
@@ -149,6 +149,14 @@ A list of authentication realm names or wildcards. The specified policy will
 
+[[xpack-sa-lf-events-ignore-privileges]]
+// tag::xpack-sa-lf-events-ignore-privileges-tag[]
+`xpack.security.audit.logfile.events.ignore_filters.<policy_name>.privileges`::
```
Personally, I think the `privileges` attribute of an `ignore_filter` is not intuitive.
I would've opted for `actions`. I acknowledge that we should generally advise for the use of
privileges instead of actions, but to me the privilege is something that the user/principal has,
whereas a policy is concerned with matching over events that contain actions.
I think a `read` action over some index name, eg:
```
xpack.security.audit.logfile.events.ignore_filters.test_policy.users: ["jack"]
xpack.security.audit.logfile.events.ignore_filters.test_policy.indices: ["meta-indices-*"]
xpack.security.audit.logfile.events.ignore_filters.test_policy.actions: ["read"]
```
is clearer.
As an extreme counter example, I think:
```
xpack.security.audit.logfile.events.ignore_filters.test_policy.users: ["jack", "ad-user-*"]
xpack.security.audit.logfile.events.ignore_filters.test_policy.privileges: ["read"]
```
conveys that the policy is about users that have the `read` privilege.

We've not really discussed this, what do you think?

albertzaharovits commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/audit/logfile/LoggingAuditTrail.java:
```diff
@@ -1355,33 +1377,40 @@ public static void registerSettings(List<Setting<?>> settings) {
+
+        private EventFilterPolicy changePrivilegesFilter(List<String> filtersList) {
+            return new EventFilterPolicy(name, ignorePrincipalsPredicate, ignoreRealmsPredicate, ignoreRolesPredicate,
+                ignoreIndicesPredicate, parsePredicate(filtersList));
```
I think some validation of the privilege names would be very welcomed!

We can use `IndexPrivilege#resolve` and `ClusterPrivilegeResolver#resolve` for every item in the list.
This raises an interesting question. There are index and cluster privileges with the same names.
For example:
```
xpack.security.audit.logfile.events.ignore_filters.test_policy.indices: ["meta-index-*"]
xpack.security.audit.logfile.events.ignore_filters.test_policy.privileges: ["monitor"]
```
ignores events for the  `monitor` actions on the "meta-index-*" namespace, but
```
xpack.security.audit.logfile.events.ignore_filters.test_policy.privileges: ["monitor"]
```
Ignores events for the monitor cluster actions and for monitor actions on all indices.

I find it a bit surprising, that because of a name clash, some, but not all, privilege names are evaluated in both a cluster and an index context.

What if 
```
xpack.security.audit.logfile.events.ignore_filters.test_policy.privileges: ["monitor"]
```
is only evaluated in a cluster context, while
```
xpack.security.audit.logfile.events.ignore_filters.test_policy.indices: ["*"]
xpack.security.audit.logfile.events.ignore_filters.test_policy.privileges: ["monitor"]
```
is the equivalent, but evaluated in an index context.
This separation would help a lot with validation.

What do you think?

BigPandaToo replied:
Added validation + tests

BigPandaToo commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into Audit_ignore_action

BigPandaToo commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into Audit_ignore_action

Commit:
To avoid ambiguity (as cluster and index policies may have the same
name) changing implementation to have to separate policies for
`index_privileges` and `cluster_privileges`.
If both are set for the same policy, throw the IllegalArgumentException.

Commit:
To avoid ambiguity (as cluster and index policies may have the same
name) changing implementation to have to separate policies for
`index_privileges` and `cluster_privileges`.
If both are set for the same policy, throw the IllegalArgumentException.

Commit:
Fixing Api key related privilege check which expects request and
authentication by introducing overloaded
version of findPrivilegesThatGrant
just checking if privileges which can grant the action regardless of the
 request and authentication context.

BigPandaToo commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into Audit_ignore_action

Commit:
Fixing a test; adding a caching mechanism to avoid calling
findPrivilegesThatGrant each
 time.

BigPandaToo commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into Audit_ignore_action

albertzaharovits commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/permission/ClusterPermission.java:
```diff
@@ -145,6 +155,14 @@ private static Automaton createAutomaton(Set<String> allowedActionPatterns, Set<
+         * @param action  action name
+         * @return {@code true} if the specified action execution can be granted by given permission else returns {@code false}
+         */
+        boolean check(String action);
```
I think I understand why you added this; you need to verify that some privileges match an action, but there are privileges (solely `manage_own_api_keys` is a named privilege like this, right?) that need to look at more than just the action to tell. In this case, the check is imperfect because the privilege might not actually grant the action. Am I right?

In case I am, I would squarely classify this as a hack. Outside of audit filtering by privileges we would never call this method.
If we're introducing a hack it should be localized (eg in `LoggingAuditTrail`). I don't like that it extends to a permission that is used in a role for authorization.

albertzaharovits commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/audit/logfile/LoggingAuditTrail.java:
```diff
@@ -1401,6 +1430,16 @@ private EventFilterPolicy changeIndicesFilter(List<String> filtersList) {
+            } else if (ClusterPrivilegeResolver.isClusterAction(action)) {
+                privileges = ClusterPrivilegeResolver.findPrivilegesThatGrant(action, null, null);
+            }
+            return privileges != null ? privileges.stream().anyMatch((s) -> ignorePrivilegesPredicate.test(s)) : false;
```
I see that you've added some caching.
But, if it is a caching from action name to privilege names, then it should be a more global one, and not at the policy level. If it is at the policy level the cache can map directly to a boolean saying if the policy matches the action.
I would say that both can work together.
If possible, make them bounded, so that if we have a bug we experience poor performance rather than grow the memory unpredictably.

albertzaharovits commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/audit/logfile/LoggingAuditTrail.java:
```diff
@@ -276,6 +289,18 @@ public LoggingAuditTrail(Settings settings, ClusterService clusterService, Threa
+            final EventFilterPolicy newPolicy = policy.orElse(new EventFilterPolicy(policyName, settings)).
+                changeClusterPrivilegesFilter(filtersList);
+            this.eventFilterPolicyRegistry.set(policyName, newPolicy);
+        }, (policyName, filtersList) -> EventFilterPolicy.parsePredicate(filtersList));
```
I see that there's still no validation per https://github.com/elastic/elasticsearch/pull/67477#discussion_r561120850 for the values in these policies.

BigPandaToo replied:
Added validation

albertzaharovits commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/audit/logfile/LoggingAuditTrail.java:
```diff
@@ -1341,12 +1374,28 @@ public static void registerSettings(List<Setting<?>> settings) {
+                    Locale.ROOT,
+                    "Both Index and Cluster privilege ignore filters are set for policy [%s]. " +
+                        "Please update your configuration settings and remove one of these filters to avoid ambiguity.", this.name);
+                throw new IllegalArgumentException(message);
```
We would need validation for the policies on node startup as well.
I remember @colings86 being a proponent of preventing the node from starting if validation fails.
This sounds reasonable, but I think it's still worth a bit of investigation and maybe socializing the direction more broadly.

In this particular case, the settings are dynamically update-able. So, if they're invalid we can set them via the persistent cluster settings API. I think it's worth going over the existing dynamic cluster settings that have validation, as well as over the settings that have validation when the node starts (I have not done so myself, maybe there's nothing interesting), and then maybe confirming the approach across the ES team.

BigPandaToo replied:
Dynamic updates which violate this rule will fail with appropriate messages now. If it happens at startup - node won't start

albertzaharovits reviewed (commented):
It's taking shape, but I see the complexity required to support privileges:

* we have actions that are authorized implicitly and are, or are not, covered by privileges
* privileges that look at more than the action name (`manage_own_api_key`)
* privileges that don't have a string representation

Granted, they're all corner cases, but they amount to blurring the association between a privilege in a policy and the events that are covered. Consumers of audit logs, which are presumably interested in trimming it down, are already exposed to actions, and I feel we're bending backwards for not much gain.

mark-vieira commented:
@elasticmachine update branch

tvernum commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into Audit_ignore_action

tvernum commented on x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/audit/logfile/LoggingAuditTrailFilterTests.java:
```diff
@@ -1851,6 +1958,88 @@ private ClusterSettings mockClusterSettings() {
+        }
+        return filtered;
+    }
+
```
I don't think the unit tests are adequate for the change we're making here.
They feel like they're mostly just adding some "action" based filters to existing tests, but this change has some significant additional requirements that need to be tested. 

Some tests that I would expect to see:
1. A filter of `cluster_privileges: all` discards all cluster events
2. A filter of `index_privileges: create_doc` works correctly (because create doc is special)
3. A filter of `cluster_privileges: manage_own_api_key` works correctly (because it is special)
4. More functional testing such as a filter of `index_privileges: index` ignores ingestion but logs deletes.

BigPandaToo replied:
Added more tests to cover all cornel cases + more functional

tvernum commented on x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/audit/logfile/LoggingAuditTrailTests.java:
```diff
@@ -342,6 +343,15 @@ public void testAuditFilterSettingValidation() {
+            containsInAnyOrder("tom", "cruise"));
+        e = expectThrows(IllegalArgumentException.class,
+            () -> LoggingAuditTrail.FILTER_POLICY_IGNORE_INDEX_PRIVILEGES.getConcreteSettingForNamespace("filter2").get(settings5));
+        assertThat(e, hasToString(containsString("invalid pattern [/foo]")));
```
As above, we need more testing around the validation.
We have made a decision that you cannot specify both index and cluster privileges, but we have no tests for that.

Commit:
Support audit ignore policy by index privileges

Addressing review feedback

BigPandaToo commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into Audit_ignore_action

BigPandaToo commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into Audit_ignore_action

BigPandaToo commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/audit/logfile/LoggingAuditTrail.java:
```diff
@@ -1410,14 +1492,17 @@ private EventFilterPolicy changeIndicesFilter(List<String> filtersList) {
             return eventInfo -> eventInfo.principal != null && ignorePrincipalsPredicate.test(eventInfo.principal)
                     && eventInfo.realm != null && ignoreRealmsPredicate.test(eventInfo.realm)
+                    && eventInfo.action != null && (ignoreIndexPrivilegesPredicateTest(eventInfo.action) &&
+                                                    ignoreClusterPrivilegesPredicateTest(eventInfo.action))
```
Changed to use permission check instead

Commit:
Support audit ignore policy by index privileges

Addressing review comments + changing approach:
- use permission check instead of simple "checkIfGrants"
- adding more testing

Commit:
Support audit ignore policy by index privileges

Addressing review comments + changing approach:
- use permission check instead of simple "checkIfGrants"
- adding more testing

Commit:
Support audit ignore policy by index privileges

Addressing review comments + changing approach:
- use permission check instead of simple "checkIfGrants"
- adding more testing

Commit:
Support audit ignore policy by index privileges

Addressing review comments + changing approach:
- use permission check instead of simple "checkIfGrants"
- adding more testing

BigPandaToo commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/AuthorizationService.java:
```diff
@@ -573,7 +573,7 @@ private static IllegalArgumentException illegalArgument(String message) {
     }
 
-    private static boolean isIndexAction(String action) {
+    public static boolean isIndexAction(String action) {
```
It is used in test

BigPandaToo commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into Audit_ignore_action

albertzaharovits commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/privilege/ManageOwnApiKeyClusterPrivilege.java:
```diff
@@ -57,6 +57,15 @@ private ManageOwnClusterPermissionCheck() {
+                throw new IllegalArgumentException(
+                    "manage own cluster permission check only supported in context of an API key request");
+            }
+
```
We should decide between returning `false` or throwing exceptions when the parameters are `null` or non-conformant, not both.

Although I would go with Exceptions because the permission in its natural authorization context shouldn't encounter such parameters (ie it's an unexpected condition), the permissions are checked in a stream and if one throws the others are not checked further, which is not what we want in the auditing context.

So my recommendation is:
```
if (request == null || authentication == null ) {
    // audit filtering may invoke this check method in a very loose way, and we shouldn't throw
    return false;
}
```

albertzaharovits commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/audit/logfile/LoggingAuditTrail.java:
```diff
@@ -1356,39 +1415,108 @@ public static void registerSettings(List<Setting<?>> settings) {
+            this.ignoreIndexPrivilegesPredicate = ignoreIndexPrivilegesPredicate;
+            this.ignoreClusterPrivilegesPredicate = ignoreClusterPrivilegesPredicate;
+
+            if (ignoreIndexPrivilegesPredicate.test("") == false && ignoreClusterPrivilegesPredicate.test("") == false) {
```
This sort of test to check if a setting is set **mostly** works but is nonetheless hard to read.
We should replace it with something like `if (ignoreIndexPrivilegesPredicate == UNSPECIFIED_PREDICATE` .

albertzaharovits commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/audit/logfile/LoggingAuditTrail.java:
```diff
@@ -1356,39 +1415,108 @@ public static void registerSettings(List<Setting<?>> settings) {
+
+            if (ignoreClusterPrivilegesPredicate.test("") == false) {
+                ClusterPermission.Builder builder = ClusterPermission.builder();
+                String[] clusterPrivileges = ignoreClusterPrivilegesPredicate.toString().split("\\|");
```
I think we should not rely on the output of `toString` for logic.

What's wrong with iterating the list of values, and then try to build the cluster permission if the value is not a wildcard or a regex? If it is a wildcard or a regex we add them to the automaton.

There's also a question here whether wildcards and regex are interpreted as matching actions and/or privileges.
I would say that wildcards (and regexps) only work for actions, i.e. `m*` is not expanded to cover all privileges that start with "m" but all actions that start with "m" (no action). I'm OK with the other option too, but we should document these.

albertzaharovits commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/audit/logfile/LoggingAuditTrail.java:
```diff
@@ -1492,9 +1637,11 @@ public String toString() {
         AuditEventMetaInfo(Optional<User> user, Optional<String> realm, Optional<AuthorizationInfo> authorizationInfo,
-                           Optional<String[]> indices) {
+                           Optional<String[]> indices, Optional<String> action, Authentication authentication,
+                           TransportRequest transportRequest) {
```
Nit: if this portion of the code uses `Optional` it's muddying to add arguments that use `null`.

albertzaharovits commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/audit/logfile/LoggingAuditTrail.java:
```diff
@@ -1356,39 +1415,108 @@ public static void registerSettings(List<Setting<?>> settings) {
+                List<IndicesPermission.Group> groups = new ArrayList<>();
+                Set<String> set = new HashSet<>(Arrays.stream(ignoreIndexPrivilegesPredicate.toString().split("\\|"))
+                    .collect(Collectors.toSet()));
+                groups.add(new IndicesPermission.Group(IndexPrivilege.get(set), FieldPermissions.DEFAULT, null, false, "*"));
```
You don't want an `IndicesPermission#Group` here. It does a lot more stuff than the checks we're after in this context (eg, data streams, backing indices, auto mapping updates).
We discussed that we don't want (can't really implement it) ignore policies for indices and index actions to work exactly like role permissions.

Instead, take the automaton from `IndexPrivilege`.

albertzaharovits reviewed (dismissed):
I did a thorough pass.
There are a couple of problems that need more consideration.

Given that we aim to have validation on the values of these filters, I think we can go without accepting `manage_own_api_key` as a valid privilege in the filter context. It's really causing a lot of headaches. This is my advice, but I've reviewed (and I'll continue to do so) as if we're supporting it.

BigPandaToo commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into Audit_ignore_action

Commit:
Revert "Support audit ignore policy by index privileges"

This reverts commit 152821e7

Commit:
Revert "Support audit ignore policy by index privileges"

This reverts commit 79649e9a

Commit:
Revert "Support audit ignore policy by index privileges"

This reverts commit 96d22a42

Commit:
Revert "Support audit ignore policy by index privileges"

This reverts commit 67574b2f

Commit:
Revert "Support audit ignore policy by index privileges"

This reverts commit 35573c8b

Commit:
Revert "Fixing a test; adding a caching mechanism to avoid calling findPrivilegesThatGrant each  time."

This reverts commit 7faa52f3

Commit:
Revert "Fixing Api key related privilege check which expects request and authentication by introducing overloaded version of findPrivilegesThatGrant just checking if privileges which can grant the action regardless of the  request and authentication context."

This reverts commit 72b9aefe

Commit:
Revert "To avoid ambiguity (as cluster and index policies may have the same name) changing implementation to have to separate policies for `index_privileges` and `cluster_privileges`. If both are set for the same policy, throw the IllegalArgumentException."

This reverts commit 7dd8fe7d

Commit:
Revert "To avoid ambiguity (as cluster and index policies may have the same name) changing implementation to have to separate policies for `index_privileges` and `cluster_privileges`. If both are set for the same policy, throw the IllegalArgumentException."

This reverts commit cb5bc09c

Commit:
Revert "Support audit ignore policy by index privileges"

This reverts commit a918da10

Commit:
Support audit ignore policy by actions

Getting back to action filtering

albertzaharovits commented on x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/audit/logfile/LoggingAuditTrailFilterTests.java:
```diff
@@ -120,52 +121,70 @@ public void testPolicyDoesNotMatchNullValuesInEvent() throws Exception {
+        Random random = random();
+        assertTrue("Matches the actions filter predicate.", auditTrail.eventFilterPolicyRegistry.ignorePredicate().test(
+            new AuditEventMetaInfo(Optional.empty(), Optional.empty(), Optional.empty(), Optional.empty(),
+                        Optional.of(filteredActions.get(random.nextInt(filteredActions.size()))))));
```
```suggestion
                        Optional.of(randomFrom(filteredActions)))));
```
There is at least one other such occurrence.

albertzaharovits commented on x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/audit/logfile/LoggingAuditTrailFilterTests.java:
```diff
@@ -1851,6 +2138,65 @@ private ClusterSettings mockClusterSettings() {
+        for (int i = 0; i < randomIntBetween(1, 4); i++) {
+            Object name = actionPatterns[random.nextInt(actionPatterns.length)];
+            filtered.add((String)name);
+        }
```
Nit you can use `randomSubsetOf(` and have a shorter list.

albertzaharovits reviewed (approved):
LGTM with only very small nits.

Commit:
Support audit ignore policy by actions

Cleaning up some tests

BigPandaToo commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into Audit_ignore_action

Commit:
Support audit ignore policy by actions

Cleaning up some tests

fbaligand commented:
Thanks for this new feature!
